include/bredis/impl/protocol.ipp

```diff
@@ -434,14 +436,58 @@ parse_result_t<Iterator, Policy> Protocol::parse(const Iterator &from,
     return details::raw_parse<Iterator, Policy>(from, to);
 }
 
-std::ostream &Protocol::serialize(std::ostream &buff,
-                                  const single_command_t &cmd) {
-    buff << '*' << (cmd.arguments.size()) << terminator;
+inline std::size_t size_for_int(std::size_t arg) {
+    std::size_t r = 0;
+    while (arg) {
+        ++r;
+        arg /= 10;
+    }
+    return std::max<size_t>(1, r); /* if r == 0 we clamp max to 1 to allow minimum containing "0" as char */
+}
+
+inline std::size_t command_size(const single_command_t &cmd) {
+    std::size_t sz = 1                                    /* * */
+                     + size_for_int(cmd.arguments.size()) /* args size */
+                     + terminator.size;
 
     for (const auto &arg : cmd.arguments) {
-        buff << '$' << arg.size() << terminator << arg << terminator;
+        sz += 1                          /* $ */
+              + size_for_int(arg.size()) /* argument size */
+              + terminator.size + arg.size() + terminator.size;
+    }
+    return sz;
+}
+
+template <typename DynamicBuffer>
+inline void Protocol::serialize(DynamicBuffer &buff,
+                                const single_command_t &cmd) {
+
+    auto it = buff.prepare(command_size(cmd));
+    constexpr std::size_t buff_sz = 64;
+    using namespace boost::asio;
+    char data[buff_sz];
+    std::size_t total =
+        snprintf(data, buff_sz, "*%zu\r\n", cmd.arguments.size());
+    buffer_copy(it, buffer(data, total));
+    it += total;
+
+    for (const auto &arg : cmd.arguments) {
+        auto bytes = snprintf(data, buff_sz, "$%zu\r\n", arg.size());
+        buffer_copy(it, buffer(data, bytes));
+        it += bytes;
+        total += bytes;
+
+        if(!arg.empty()) {
+            buffer_copy(it, buffer(arg.data(), arg.size()));
+            it += arg.size();
+            total += arg.size();
+        }
+
+        buffer_copy(it, buffer("\r\n", terminator.size));
+        total += terminator.size;
+        it += terminator.size;
     }
-    return buff;
+    buff.commit(total);
```

Comment:
I think this function is designed to just work with **boost::asio::streambu**. so what is the point of using template in its definition?